framework/i18n/Formatter.php

```diff
@@ -1741,4 +1778,220 @@ protected function createNumberFormatter($style, $decimals = null, $options = []
+    }
+
+    /**
+     * Fallback for formatting value as a decimal number.
+     *
+     * Property [[decimalSeparator]] will be used to represent the decimal point. The value is rounded automatically
+     * to the defined decimal digits.
+     *
+     * @param string|int|float $value the value to be formatted.
+     * @param int $decimals the number of digits after the decimal point. The default value is `2`.
+     * @return string the formatted result.
+     * @see decimalSeparator
+     * @see thousandSeparator
+     * @since 2.0.16
+     */
+    protected function asDecimalStringFallback($value, $decimals = null)
+    {
+        if (empty($value)) {
+            $value = 0;
+        }
+
+        $value = (string) $value;
+
+        if (strpos($value, '.') !== false) {
+            $integerPart = substr($value, 0, strrpos($value, '.'));
+            $fractionalPart = substr($value, strrpos($value, '.') + 1);
+        } else {
+            $integerPart = $value;
+            $fractionalPart = null;
+        }
+
+        $decimalOutput = '';
+
+        if ($decimals === null) {
+            $decimals = 2;
+        }
+
+        $roundIntegerPart = false;
+
+        if ($decimals > 0) {
+            $decimalSeparator = $this->decimalSeparator;
+            if ($this->decimalSeparator === null) {
+                $decimalSeparator = '.';
+            }
+
+            if ($fractionalPart === null) {
+                $fractionalPart = str_repeat('0', $decimals);
+            } elseif (strlen($fractionalPart) > $decimals) {
+                $cursor = $decimals;
+
+                if ((int) substr($fractionalPart, $cursor, 1) >= 5) {
+                    while (--$cursor >= 0) {
+                        $round = false;
+                        $newDigit = (int) substr($fractionalPart, $cursor, 1) + 1;
+
+                        if ($newDigit === 10) {
+                            $newDigit = 0;
+                            $round = true;
+                        }
+
+                        $fractionalPart = substr($fractionalPart, 0, $cursor) . $newDigit . substr($fractionalPart, $cursor + 1);
+
+                        if (!$round) {
+                            break;
+                        }
+                    }
+                    if ($round) {
+                        $roundIntegerPart = true;
+                    }
+                }
+
+                $fractionalPart = substr($fractionalPart, 0, $decimals);
+            } elseif (strlen($fractionalPart) < $decimals) {
+                $fractionalPart = str_pad($fractionalPart, $decimals, '0');
+            }
+
+            $decimalOutput .= $decimalSeparator . $fractionalPart;
+        }
+
+        if ($roundIntegerPart || ($decimals === 0 && $fractionalPart !== null && (int) substr($fractionalPart, 0, 1) >= 5)) {
+            $cursor = 0;
+            while (++$cursor <= strlen($integerPart)) {
+                $round = false;
```

Comment:
That looks like addition is happening. Worth using `$carry`.

Reply:
Also done.